From 14a420c26797f4350af69a31c880b416ad09e2e5 Mon Sep 17 00:00:00 2001
From: Hao Sun <hao.sun@inf.ethz.ch>
Date: Wed, 5 Nov 2025 11:00:47 +0100
Subject: [PATCH RFC 7/7] bpf: Temporarily treat unsupported as trusted

Add apply_trusted_step() and use it for two cases: REWRITE with UNSPEC id
(single argument) and POLY_NORM. This is for debug purpose only, and the
rules will bed added in followups.

Signed-off-by: Hao Sun <hao.sun@inf.ethz.ch>
---
 kernel/bpf/bcf_checker.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/kernel/bpf/bcf_checker.c b/kernel/bpf/bcf_checker.c
index d4105ff957a0..dbea3df2ba87 100644
--- a/kernel/bpf/bcf_checker.c
+++ b/kernel/bpf/bcf_checker.c
@@ -1452,6 +1452,13 @@ static int set_step_fact_id(struct bcf_checker_state *st, u32 fact_id)
 	return 0;
 }
 
+static int apply_trusted_step(struct bcf_checker_state *st, char *rule_name,
+			      u32 fact_id)
+{
+	verbose(st, "; WARNING: applying trusted step %s\n", rule_name);
+	return set_step_fact_id(st, fact_id);
+}
+
 static struct bcf_expr *get_premise(struct bcf_checker_state *st,
 				    struct bcf_proof_step *step, u32 arg)
 {
@@ -2698,6 +2705,13 @@ REWRITE: { /* Rewrite equality to equivalent expression */
 	rewrite_id = step->args[pm_step_n];
 	args = step->args + step->premise_cnt + 1;
 	arg_n = step->param_cnt - 1;
+
+	if (rewrite_id == BCF_REWRITE_UNSPEC) {
+		if (arg_n == 1 && !pm_step_n)
+			return apply_trusted_step(st, "rewrite", args[0]);
+		return -EINVAL;
+	}
+
 	err = apply_rewrite(st, &fact, rewrite_id, pm_steps, pm_step_n, args,
 			    arg_n);
 	if (err)
@@ -4624,7 +4638,7 @@ POLY_NORM: { /* Equality of polynomial normal form */
 	if (IS_ERR(bv_eq))
 		return PTR_ERR(bv_eq);
 	ENSURE(is_bool_eq(bv_eq->code));
-	return -EOPNOTSUPP;
+	return apply_trusted_step(st, "POLY_NORM", step->args[0]);
 }
 
 POLY_NORM_EQ: { /* Polynomial normalization for relations */
-- 
2.34.1

